### DISCUSSION TITLE
Docs on plugins for building extension modules?

### DISCUSSION THREAD

**Discussion by codeandfire** [pypa/setuptools#3458] [Q&A] [answered]
Hi,

From the docs on [building extension modules](https://setuptools.pypa.io/en/latest/userguide/ext_modules.html), I can see that Setuptools has in-built functionality for building C/C++ extension modules, and Cython extensions can be built with an extra build-time `cython` dependency. For other tools like CMake or other languages like Rust and Go, I believe the solution is to write a plugin that overrides the Setuptools `build_ext` command. However, I can't find much documentation on writing such plugins: [this page](https://setuptools.pypa.io/en/latest/userguide/extension.html) in the Setuptools docs doesn't cover the `build_ext` command, and the Distutils docs ([here](https://docs.python.org/3/distutils/extending.html), [here](https://docs.python.org/3/distutils/examples.html#single-extension-module) and [here](https://docs.python.org/3/distutils/apiref.html#module-distutils.command.build_ext)) don't seem to cover them either.

Is there any place where I can find documentation on this?

P.S. I know that there are existing plugins for Rust and Go and possibly even CMake, and that I probably do not have write them myself. But I still want to understand how this really works, hence the question.

**Comment by abravalheri** [comment, marked as answer]
Sorry for the delay in replying @codeandfire.

> Is there any place where I can find documentation on this?

I am afraid that the official documentation still does not cover that.

The following doc may be related https://setuptools--3485.org.readthedocs.build/en/3485/userguide/ext_modules.html

In preparation for the upcoming PEP 660 implementation I would also advise about having a look on https://setuptools--3485.org.readthedocs.build/en/3485/userguide/extension.html#customizing-commands.

In terms of implementation, I would take advantage of the `setuptools.finalize_distribution_options` entry-point to modify the `cmdclass` attribute of the distribution object and, this way, inject a custom `build_ext` or define a separated custom build step.

**Reply by codeandfire** [reply]
Cool, thank you very much!
